src/huggingface_inference_toolkit/handler.py

```diff
@@ -12,34 +14,86 @@
-            model_dir=model_dir,
-            task=task,
+            model_dir=model_dir,  # type: ignore
+            task=task,  # type: ignore
             framework=framework,
             trust_remote_code=HF_TRUST_REMOTE_CODE,
         )
 
-    def __call__(self, data):
+    def __call__(self, data: Dict[str, Any]) -> Dict[str, Any]:
         """
         Handles an inference request with input data and makes a prediction.
         Args:
             :data: (obj): the raw request body data.
         :return: prediction output
         """
         inputs = data.pop("inputs", data)
-        parameters = data.pop("parameters", None)
-
-        # pass inputs with all kwargs in data
-        if parameters is not None:
-            prediction = self.pipeline(inputs, **parameters)
-        else:
-            prediction = self.pipeline(inputs)
-        # postprocess the prediction
-        return prediction
+        parameters = data.pop("parameters", {})
+
+        # sentence transformers pipelines do not have the `task` arg
+        if any(isinstance(self.pipeline, v) for v in SENTENCE_TRANSFORMERS_TASKS.values()):
+            return self.pipeline(**inputs) if isinstance(inputs, dict) else self.pipeline(inputs)  # type: ignore
+
+        if self.pipeline.task == "question-answering":
+            if not isinstance(inputs, dict):
+                raise ValueError(f"inputs must be a dict, but a `{type(inputs)}` was provided instead.")
+            if not all(k in inputs for k in {"question", "context"}):
+                raise ValueError(
+                    f"{self.pipeline.task} expects `inputs` to be a dict containing both `question` and "
+                    "`context` as the keys, both of them being either a `str` or a `List[str]`."
+                )
+
+        if self.pipeline.task == "table-question-answering":
+            if not isinstance(inputs, dict):
+                raise ValueError(f"inputs must be a dict, but a `{type(inputs)}` was provided instead.")
+            if "question" in inputs:
+                inputs["query"] = inputs.pop("question")
+            if not all(k in inputs for k in {"table", "query"}):
+                raise ValueError(
+                    f"{self.pipeline.task} expects `inputs` to be a dict containing the keys `table` and "
+                    "either `question` or `query`."
+                )
+
+        if self.pipeline.task in {"token-classification", "ner"}:
+            # stride and aggregation_strategy are defined on `pipeline` init, but in the Inference API those
+            # are provided on each request instead
+            for p in {"stride", "aggregation_strategy"}:
+                if p in parameters:
+                    parameters.pop(p)
+                    logger.warning(f"provided parameter `{p}`, but it's not supported.")
+
+        if self.pipeline.task.__contains__("translation"):
+            # truncation and generate_parameters are used on Inference API but not available on
```

Comment:
`TranslationPipeline` inherits parameter handling from its parent class, see [here](https://github.com/huggingface/transformers/blob/052e652d6d53c2b26ffde87e039b723949a53493/src/transformers/pipelines/text2text_generation.py#L331). This means it uses both its own parameters and those defined in the parent `Text2TextGenerationPipeline` at [transformers/pipelines/text2text_generation.py#L75](https://github.com/huggingface/transformers/blob/052e652d6d53c2b26ffde87e039b723949a53493/src/transformers/pipelines/text2text_generation.py#L75). Regarding "generate_parameters", this is supposed to replace generate_kwargs in the future, more context [here](https://github.com/huggingface/huggingface.js/issues/923).

Reply:
Thanks for the note here @hanouticelina!

As of `generate_parameters`, it seems that the default on Transformers is still `generate_kwargs` even though it will be replaced by `generation_paramters` instead of `generate_parameters`, right? As we're pinning the Transformers version, we're good on renaming those back to `generate_kwargs` as it's the one currently supported, while still being able to handle both `generate_parameters` and `generation_parameters` to come on upcoming releases, is that right? (see the udpated code in this same section for it)

Reply:
looking at the code in `transformers`, I realized that `generate_kwargs` for `translation` and `text-generation` tasks  is a mapping of keyword arguments (see [here](https://github.com/huggingface/transformers/blob/v4.46.3/src/transformers/pipelines/text2text_generation.py#L83) and [here](https://github.com/huggingface/transformers/blob/v4.46.3/src/transformers/pipelines/text_generation.py#L135)), which is not the case for other tasks, eg: [image-to-text](https://github.com/huggingface/transformers/blob/v4.46.3/src/transformers/pipelines/image_to_text.py#L77). that means you don't need to pass `generate_kwargs` (or `generation_parameters`) as a dictionary — the generation parameters should be passed directly to the pipeline call:

this throws an error:
```python
from transformers import pipeline

pipe = pipeline("translation", model="facebook/mbart-large-50-many-to-many-mmt")

output = pipe("bonjour", src_lang="fr", tgt_lang="en", generate_kwargs={"max_new_tokens": 100})
print(output)
# throws : ValueError: The following `model_kwargs` are not used by the model: ['generate_kwargs'] (note: typos in the generate arguments will also show up in this list)
```
but this works fine:
```python
from transformers import pipeline

pipe = pipeline("translation", model="facebook/mbart-large-50-many-to-many-mmt")

output = pipe("bonjour", src_lang="fr", tgt_lang="en", max_new_tokens=100)
print(output)
```
i will fix this also in the translation [specs](https://github.com/huggingface/huggingface.js/blob/main/packages/tasks/src/tasks/translation/spec/input.json) in `huggingface.js` and in the `huggingface_hub` client!

Reply:
Oh great nice catch indeed, let me update it here too! 👍🏻